### ISSUE TITLE
[JENKINS-60875] branch index scan of multibranch pipeline job fails with GitException on old CLI git (1.8.3)

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/git-plugin#3513]
<p>A multibranch pipeline jobs is configured for a git repo. Scanning the branch index ends with an excpetion from git client:</p>

<pre>
Mon Jan 27 09:21:00 CET 2020] Starting branch indexing...
 > git --version # timeout=10
using GIT_ASKPASS to set credentials User <span class="code-keyword">for</span> XXX bitbucket instance
Setting http proxy: XXX.XXX.XXX.X:9090
 > git ls-remote https:<span class="code-comment">//bitbucket.xxx.yy/somewhat.git # timeout=10
</span>ERROR: [Mon Jan 27 09:21:01 CET 2020] Could not update folder level actions from source <uuid ommited> 
hudson.plugins.git.GitException: Command <span class="code-quote">"git ls-remote https:<span class="code-comment">//bitbucket.xxx.yy/somewhat.git"</span> returned status code 128:
</span>stdout: 
stderr: fatal: unable to access <span class="code-quote">'.git/config'</span>: Permission denied

	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2372)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:1985)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:1885)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:1876)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.getRemoteReferences(CliGitAPIImpl.java:3228)
	at jenkins.plugins.git.AbstractGitSCMSource.retrieveActions(AbstractGitSCMSource.java:1141)
	at jenkins.scm.api.SCMSource.fetchActions(SCMSource.java:848)
	at jenkins.branch.MultiBranchProject.computeChildren(MultiBranchProject.java:598)
	at com.cloudbees.hudson.plugins.folder.computed.ComputedFolder.updateChildren(ComputedFolder.java:277)
	at com.cloudbees.hudson.plugins.folder.computed.FolderComputation.run(FolderComputation.java:164)
	at jenkins.branch.MultiBranchProject$BranchIndexing.run(MultiBranchProject.java:1032)
	at hudson.model.ResourceController.execute(ResourceController.java:97)
	at hudson.model.Executor.run(Executor.java:428)
[Mon Jan 27 09:21:01 CET 2020] Finished branch indexing. Indexing took 91 ms</pre>
<p>It seems, that git client plugin tries to run the <tt>git ls-remote</tt> command somewhere, where the jenkins instance does not have write access to.</p>

<p>By looking into the code, I saw that</p>

<pre>
org.jenkinsci.plugins.gitclient.CliGitAPIImpl.getRemoteReferences
</pre>
<p>does not define a workdir for the git command call and so the process default workdir is chosen.</p>

<p>The pwd of the java vm running jenkins is <tt>/</tt></p>

<p>That would explain the problem.</p>

<p>Idea: Use <tt>java.io.tmpdir</tt> to get a good workdir, if nothing else fits.</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=gerrietr">gerrietr</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-60875" target="_blank">branch index scan of multibranch pipeline job fails with GitException on old CLI git (1.8.3)</a></i></summary>
<i><ul>
<li><b>status</b>: Closed
<li><b>priority</b>: Minor
<li><b>component(s)</b>: git-plugin
<li><b>resolution</b>: Won't Fix
<li><b>resolved</b>: 2020-07-12T00:57:00+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 2
<li><b>imported</b>: 2025-12-02
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>A multibranch pipeline jobs is configured for a git repo. Scanning the branch index ends with an excpetion from git client:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">Mon Jan 27 09:21:00 CET 2020] Starting branch indexing...
 &gt; git --version # timeout=10
using GIT_ASKPASS to set credentials User <span class="code-keyword">for</span> XXX bitbucket instance
Setting http proxy: XXX.XXX.XXX.X:9090
 &gt; git ls-remote https:<span class="code-comment">//bitbucket.xxx.yy/somewhat.git # timeout=10
</span>ERROR: [Mon Jan 27 09:21:01 CET 2020] Could not update folder level actions from source &lt;uuid ommited&gt; 
hudson.plugins.git.GitException: Command <span class="code-quote">"git ls-remote https:<span class="code-comment">//bitbucket.xxx.yy/somewhat.git"</span> returned status code 128:
</span>stdout: 
stderr: fatal: unable to access <span class="code-quote">'.git/config'</span>: Permission denied

	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2372)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:1985)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:1885)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:1876)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.getRemoteReferences(CliGitAPIImpl.java:3228)
	at jenkins.plugins.git.AbstractGitSCMSource.retrieveActions(AbstractGitSCMSource.java:1141)
	at jenkins.scm.api.SCMSource.fetchActions(SCMSource.java:848)
	at jenkins.branch.MultiBranchProject.computeChildren(MultiBranchProject.java:598)
	at com.cloudbees.hudson.plugins.folder.computed.ComputedFolder.updateChildren(ComputedFolder.java:277)
	at com.cloudbees.hudson.plugins.folder.computed.FolderComputation.run(FolderComputation.java:164)
	at jenkins.branch.MultiBranchProject$BranchIndexing.run(MultiBranchProject.java:1032)
	at hudson.model.ResourceController.execute(ResourceController.java:97)
	at hudson.model.Executor.run(Executor.java:428)
[Mon Jan 27 09:21:01 CET 2020] Finished branch indexing. Indexing took 91 ms</pre>
</div></div>
<p>It seems, that git client plugin tries to run the <tt>git ls-remote</tt> command somewhere, where the jenkins instance does not have write access to.</p>

<p>By looking into the code, I saw that</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">org.jenkinsci.plugins.gitclient.CliGitAPIImpl.getRemoteReferences
</pre>
</div></div>
<p>does not define a workdir for the git command call and so the process default workdir is chosen.</p>

<p>The pwd of the java vm running jenkins is <tt>/</tt></p>

<p>That would explain the problem.</p>

<p>Idea: Use <tt>java.io.tmpdir</tt> to get a good workdir, if nothing else fits.</p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
os: CentOS 7<br/>
jdk: openjdk-1.8.0.191<br/>
Jenkins version 2.210<br/>
git client: 3.0.0<br/>
```
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-60875?focusedId=383788&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-383788">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>What version of command line git are you running on that computer?  The <tt>git ls-remote</tt> command should not attempt to write to the <tt>.git/config</tt> file and should gracefully handle attempts to read from the file.</p>

<p>The issue mentions CentOS 7.  The default version of command line git on CentOS 7 is git 1.8.  That version of command line git is several years behind the current command line git 2.25.0.</p>

<p>As a work around, you could set the working directory of the process that starts Jenkins to the JENKINS_HOME directory before starting the Jenkins process. </p>

<p>Scanning the remote repository with <tt>git ls-remote</tt> is generally expected to be performed from a local copy of the repository.  Setting the working directory inside the <tt>getRemoteReferences</tt> call seems like that risks overriding the working directory as set by other processes.</p>
</pre>
</details>

</ul>
</details>

<p>What version of command line git are you running on that computer?  The <tt>git ls-remote</tt> command should not attempt to write to the <tt>.git/config</tt> file and should gracefully handle attempts to read from the file.</p>

<p>The issue mentions CentOS 7.  The default version of command line git on CentOS 7 is git 1.8.  That version of command line git is several years behind the current command line git 2.25.0.</p>

<p>As a work around, you could set the working directory of the process that starts Jenkins to the JENKINS_HOME directory before starting the Jenkins process. </p>

<p>Scanning the remote repository with <tt>git ls-remote</tt> is generally expected to be performed from a local copy of the repository.  Setting the working directory inside the <tt>getRemoteReferences</tt> call seems like that risks overriding the working directory as set by other processes.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>gerrietr:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-60875?focusedId=383794&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-383794">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Thanks for your quick reply!</p>

<p>Indeed, local git version is 1.8.3.1. I will try to change the working directory at startup of jenkins.</p>

<p>But maybe, there is an issue with multibranch jobs in general, here. When scanning for branches, there seems to be <b>no</b> local copy of the repository!?</p>
</pre>
</details>

</ul>
</details>

<p>Thanks for your quick reply!</p>

<p>Indeed, local git version is 1.8.3.1. I will try to change the working directory at startup of jenkins.</p>

<p>But maybe, there is an issue with multibranch jobs in general, here. When scanning for branches, there seems to be <b>no</b> local copy of the repository!?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-60875?focusedId=383795&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-383795">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Branch source plugins that use a REST API call to request the changes from the remote repository do not need a local repository.  Since you report that you're using Bitbucket, I assume you're using the Bitbucket branch source plugin.  I thought it was modeled after the GitHub and Gitea plugins which use REST API calls rather than using <tt>git ls-remote</tt>.</p>

<p>If you're using the git plugin to provide the branch source (git as branch source provider instead of Bitbucket), then you'll usually find the cached repositories in the <tt>$JENKINS_HOME/caches</tt> directory.</p>
</pre>
</details>

</ul>
</details>

<p>Branch source plugins that use a REST API call to request the changes from the remote repository do not need a local repository.  Since you report that you're using Bitbucket, I assume you're using the Bitbucket branch source plugin.  I thought it was modeled after the GitHub and Gitea plugins which use REST API calls rather than using <tt>git ls-remote</tt>.</p>

<p>If you're using the git plugin to provide the branch source (git as branch source provider instead of Bitbucket), then you'll usually find the cached repositories in the <tt>$JENKINS_HOME/caches</tt> directory.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>gerrietr:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-60875?focusedId=383796&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-383796">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I haved used the git plugin. Now i have tried the Bitbucket branch source plugin. There everything works fine!</p>

<p>Thanks for the hint!</p>
</pre>
</details>

</ul>
</details>

<p>I haved used the git plugin. Now i have tried the Bitbucket branch source plugin. There everything works fine!</p>

<p>Thanks for the hint!</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-60875?focusedId=383797&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-383797">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>The problem code path in <a href="https://github.com/jenkinsci/git-plugin/blob/16ef0dafd78a5573eb6cf2c4a11d64910f2e3765/src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java#L1141" class="external-link" target="_blank" rel="nofollow noopener">AbstractGitSCMSource</a> is only reachable with command line git versions older than git 2.8.0.  That includes CentOS 6, CentOS 7, Debian 7, Debian 8, Ubuntu 14, and Ubuntu 16 in their default configurations.  </p>

<p>CentOS 8, Ubuntu 18, Debian 9, and Debian 10 all include newer versions of command line git which avoid that code path and use the symref feature of newer command line git versions.</p>
</pre>
</details>

</ul>
</details>

<p>The problem code path in <a href="https://github.com/jenkinsci/git-plugin/blob/16ef0dafd78a5573eb6cf2c4a11d64910f2e3765/src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java#L1141" class="external-link" target="_blank" rel="nofollow noopener">AbstractGitSCMSource</a> is only reachable with command line git versions older than git 2.8.0.  That includes CentOS 6, CentOS 7, Debian 7, Debian 8, Ubuntu 14, and Ubuntu 16 in their default configurations.  </p>

<p>CentOS 8, Ubuntu 18, Debian 9, and Debian 10 all include newer versions of command line git which avoid that code path and use the symref feature of newer command line git versions.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>gerrietr:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-60875?focusedId=385737&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-385737">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>The same problem occurs when I configure a Global Pipeline Library with git as source.</p>
</pre>
</details>

</ul>
</details>

<p>The same problem occurs when I configure a Global Pipeline Library with git as source.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-60875?focusedId=394101&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-394101">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I don't intend to work on this because command line git versions older than 2.8.0 are increasingly uncommon and alternatives exist that will allow users to install newer versions even on operating systems like CentOS 7.</p>
</pre>
</details>

</ul>
</details>

<p>I don't intend to work on this because command line git versions older than 2.8.0 are increasingly uncommon and alternatives exist that will allow users to install newer versions even on operating systems like CentOS 7.</p>
